src/main/java/com/aws/greengrass/clientdevices/auth/connectivity/CISShadowMonitor.java

```diff
@@ -195,35 +217,98 @@ private void subscribeToShadowTopics() throws InterruptedException {
+                    .exceptionally(e -> {
+                        LOGGER.atWarn().cause(e).log("Unable to update CIS shadow reported state");
+                        return null;
+                    });
+        } catch (Exception e) {
+            LOGGER.atWarn().cause(e).log("Unable to send PublishUpdateShadow request");
         }
+    }
 
-        if (Objects.equals(version, lastVersion)) {
-            LOGGER.atInfo().kv(VERSION, version).log("Already processed version. Skipping cert re-generation");
-            updateCISShadowReportedState(desiredState);
-            return;
+    private void publishToGetCISShadowTopic() {
+        LOGGER.atDebug().log("Publishing to get shadow topic");
+        GetShadowRequest getShadowRequest = new GetShadowRequest();
+        getShadowRequest.thingName = shadowName;
+        iotShadowClient.PublishGetShadow(getShadowRequest, QualityOfService.AT_LEAST_ONCE).exceptionally(e -> {
+            LOGGER.atWarn().cause(e).log("Unable to retrieve CIS shadow");
+            return null;
+        });
+    }
+
+    private void unsubscribeFromShadowTopics() {
+        if (connection != null) {
+            LOGGER.atDebug().log("Unsubscribing from CIS shadow topics");
+            String topic = String.format(SHADOW_UPDATE_DELTA_TOPIC, shadowName);
+            connection.unsubscribe(topic);
+
+            topic = String.format(SHADOW_GET_ACCEPTED_TOPIC, shadowName);
+            connection.unsubscribe(topic);
         }
+    }
 
-        LOGGER.atInfo().log("New CIS version: {}", version);
+    @Override
+    public void accept(NetworkStateProvider.ConnectionState state) {
+        if (state == NetworkStateProvider.ConnectionState.NETWORK_UP) {
+            publishToGetCISShadowTopic();
+        }
+    }
 
-        // NOTE: This method executes in an MQTT CRT thread. Since certificate generation is a compute intensive
-        // operation (particularly on low end devices) it is imperative that we process this event asynchronously
-        // to avoid blocking other MQTT subscribers in the Nucleus
-        CompletableFuture.runAsync(() -> {
+    /**
+     * Task for processing the CIS shadow. The goal is to rotate
+     * certificates if we detect that connectivity info has changed.
+     * This task is triggered by
+     */
+    private class ProcessCISShadowTask implements Runnable {
 
+        @Getter
+        private final int shadowVersion;
+        private final Map<String, Object> desiredState;
+
+        ProcessCISShadowTask(int shadowVersion, Map<String, Object> desiredState) {
+            this.shadowVersion = shadowVersion;
+            this.desiredState = desiredState;
+        }
+
+        @Override
+        @SuppressWarnings("PMD.AvoidCatchingGenericException")
+        public void run() {
+            try {
+                processCISShadow();
+            } catch (InterruptedException ignored) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)